sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,488 @@
+defines kernels and these kernels are run by the application unchanged cache
+structure may lead to double compilation of the same code. Example:
+```
+// Application
+SYCL_EXTERNAL void LibFunc();
+
+Q.submit([&](cl::sycl::handler &CGH) {
+handler.parallel_for([] { LibFunc(); });  // Device image for library is compiled
+                                          // and linked together with device
+                                          // image for application, i.e.
+                                          // LibFunc1 and ExternalKernel exist
+                                          // in prepared state
+});
+// ...
+EnqueueLibraryKernel(Q); // If cache mechanism is not changed, this line will
+                         // lead to second compilation of ExternalKernel and
+                         // LibFunc1
+
+// Library
+SYCL_EXTERNAL void LibFunc1() {
+// ...
+}
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+```
+Such case can be optimized by bringing nesting into cache keys structure.
+Kernel set id can be found for each kernel using its name and OS module it is
+coming from. In presence of dynamic linking resulting program can be combined
+out of device images which come from different OS modules. So, it should be
+possible to find needed program by kernel name and any OS module that was
+involved in this program. The new mapping structure is:
+```
+{kernel name} =>
+  {OSModuleHandle, spec const, opts, dev} => program
+```
+I.e. each kernel name is mapped to a set of tuples that consists of OS module,
+spec constant values, JIT compiler options, device. Then concrete tuple is
+mapped to a program object.
+Example:
+```
+// Application
+// OSModule = 1
+
+SYCL_EXTERNAL void LibFunc();
+queue.submit(parallel_for<InternalKernel>( ... ));
+
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([] { LibFunc(); });
+});
+
+EnqueueLibraryKernel(q);
+
+// Library
+// OSModule = 2
+
+SYCL_EXTERNAL lib1_func();
+
+EnqueueLibraryKernel(queue) {
+  queue.submit(parallel_for<ExternalKernel>(...));
+}
+
+Program cache will have the following structure:
+ "InternalKernel" =>
+   {1, ...} => program 1
+ "ExternalKernel" =>
+   {1, ...} => program 1
+   {2, ...} => program 1
+```
+However the library code will be compiled twice if kernel from the library
+was enqueued before kernels from the application, i.e. in such case:
+```
+// Application
+SYCL_EXTERNAL void LibFunc();
+
+EnqueueLibraryKernel(Q); // First, library code is compiled alone since it
+                         // doesn't have any dependencies
+// ...
+Q.submit([&](cl::sycl::handler &CGH) {
+handler.parallel_for([] { LibFunc(); });  // Second, library code is compiled
+                                          // and linked together with code of
+                                          // the application
+});
+```
+
+The program caching mechanism is re-used without changes.
```

Comment:
The program caching mechanism was changed in the above text.
Should this sentence say about kernel cache structure?

Reply:
Yes, it should say about kernel cache. Will fix that.